src/contract/event.rs

```diff
@@ -293,6 +285,196 @@ impl<T: Transport, E: Detokenize> Stream for EventStream<T, E> {
+    pub topics: Vec<H256>,
+    /// The raw non-indexed data attached to an event.
+    pub data: Vec<u8>,
+}
+
+impl RawLog {
+    /// Decode raw log data into a tokenizable for a matching event ABI entry.
+    pub fn decode<D>(self, event: &AbiEvent) -> Result<D, ExecutionError>
+    where
+        D: Detokenize,
+    {
+        let event_log = event.parse_log(AbiRawLog {
+            topics: self.topics,
+            data: self.data,
+        })?;
+
+        let tokens = event_log
+            .params
+            .into_iter()
+            .map(|param| param.value)
+            .collect::<Vec<_>>()
+            .compat()
+            .ok_or(ExecutionError::UnsupportedToken)?;
+        let data = D::from_tokens(tokens)?;
+
+        Ok(data)
+    }
+}
+
+impl From<Log> for RawLog {
+    fn from(log: Log) -> Self {
+        RawLog {
+            topics: log.topics,
+            data: log.data.0,
+        }
+    }
+}
+
+impl ParseLog for RawLog {
+    fn parse_log(log: RawLog) -> Result<Self, ExecutionError> {
+        Ok(log)
+    }
+}
+
+/// A builder for creating a filtered stream for any contract event.
+#[must_use = "event builders do nothing unless you stream them"]
+pub struct AllEventsBuilder<T: Transport, E: ParseLog> {
+    /// The underlying web3 instance.
+    web3: Web3<T>,
+    /// The web3 filter builder used for creating a log filter.
+    filter: FilterBuilder,
+    /// The topic filters that are encoded based on the event ABI.
+    pub topics: TopicFilter,
+    /// The polling interval for querying the node for more events.
+    pub poll_interval: Option<Duration>,
+    _events: PhantomData<E>,
+}
+
+impl<T: Transport, E: ParseLog> AllEventsBuilder<T, E> {
+    /// Creates a new all events builder from a web3 provider and and address.
+    pub fn new(web3: Web3<T>, address: Address) -> Self {
+        AllEventsBuilder {
+            web3,
+            filter: FilterBuilder::default().address(vec![address]),
+            topics: TopicFilter::default(),
+            poll_interval: None,
+            _events: PhantomData,
+        }
+    }
+
+    /// Sets the starting block from which to stream logs for.
+    ///
+    /// If left unset defaults to the latest block.
+    #[allow(clippy::wrong_self_convention)]
+    pub fn from_block(mut self, block: BlockNumber) -> Self {
+        self.filter = self.filter.from_block(block);
+        self
+    }
+
+    /// Sets the last block from which to stream logs for.
+    ///
+    /// If left unset defaults to the streaming until the end of days.
+    #[allow(clippy::wrong_self_convention)]
+    pub fn to_block(mut self, block: BlockNumber) -> Self {
+        self.filter = self.filter.to_block(block);
+        self
+    }
+
+    /// Adds a filter for the first indexed topic.
+    ///
+    /// For regular events, this corresponds to the event signature. For
+    /// anonymous events, this is the first indexed property.
+    pub fn topic0(mut self, topic: Topic<H256>) -> Self {
```

Comment:
You can filter on all_events? How does that work? You specify the H256 hash of the thing you want to filter on?

Reply:
You manually specify the topics you want to filter on. This can mean different things depending on the event being filtered.

Reply:
The reason this has a different signature than the `EventBuilder` type is that, for that type, it makes sense to handle conversion of a topic with a single data type. This is because the first indexed property given an event signature has a known type, and so specifying a `Topic::OneOf(Vec<SomeAbiType>)` filter makes sense.

For `AllEventsBuilder` this is no longer the case as you may be building a filter for various events where the first indexed property no longer has the same type. In this case the `Topic::OneOf` filter needs to be contain the topic values derived from different types and can no longer be specified from a `Vec<SomeSingleType>` but needs to be a `Vec<EnumWithPossibleAllTypesForThisTopic>`.

Earlier this week I tried really hard to get this working with some type arithmetic, in a way to make this very type-safe for the generated code, but I wasn't able to figure it out in a nice way. The hope was to get something like this for the generated code:

```
instance.all_events()
    .event::<Transfer>()
    .event::<Approval>()
    .topic0(/* here only accepts `Address` or `Vec<Address>` since that is the common first indexed property for both events */)
    .stream()
```

I will maybe revisit this later, but this would have been way more complex than how it is currently implemented and I couldn't quite get it working.

Reply:
Thanks for the explanation, this makes sense. Is there a way to derive the H256 for let's say a set of concrete EventBuilder topic filters?

Reply:
Yes, currently, we already convert `Tokenizable` types into these values by validating the token kinds against the event entry in the ABI, we also want to generate `EventBuilder` types for each event kind so that the interaction looks like this
```
instance.events()
    .transfer()
    .from(some_single_address)
    .to(&[some, addresses, to, filter])
    .stream()
```

But that needs to be in a separate PR.